README.md
# batchproc
[![Go Reference](https://pkg.go.dev/badge/github.com/aryannr97/batchproc.svg)](https://pkg.go.dev/github.com/aryannr97/batchproc)
[![RELEASE](https://github.com/aryannr97/batchproc/actions/workflows/release.yml/badge.svg)](https://github.com/aryannr97/batchproc/actions/workflows/release.yml)
![GitHub go.mod Go version](https://img.shields.io/github/go-mod/go-version/aryannr97/batchproc)
[![Codecov](https://img.shields.io/codecov/c/github/aryannr97/batchproc)](https://app.codecov.io/gh/aryannr97/batchproc)
[![Go Report Card](https://goreportcard.com/badge/github.com/aryannr97/batchproc)](https://goreportcard.com/report/github.com/aryannr97/batchproc)
[![Go Report Card](https://img.shields.io/badge/Linter-golangci--lint-informational)](https://golangci-lint.run)
[![MIT license](https://img.shields.io/github/license/aryannr97/batchproc)](https://github.com/aryannr97/batchproc/blob/main/LICENSE)
## Generic batch processor

batchproc is developed to serve the purpose of enabling a uniform way to process large volumes of different types of data & run repititive computations as per requirement 

batchproc essentially provides
- Ability to process any type of **indexed**, **iterative** data
- **Customization** for batch steps
- [**Rungroups**](https://github.com/bharat-rajani/rungroup) to execute batch units concurrently

### :floppy_disk: Installation
```
go get -u github.com/aryannr97/batchproc
```

### :notebook_with_decorative_cover: Documentation
Detailed documentation can be found [here](https://github.com/aryannr97/batchproc/wiki)

### :technologist: Usage

##### 1. **New** 
- Returns a new instance of batch processing executor.
```go
// Creation stage
executor := batchproc.New(context.Background(), "unique-id", len(data), data, getBatchUnit)
```
- **GetBatchUnitFunc** is a predefined type `type GetBatchUnitFunc func() BatchUnit`
- User should define & pass this function while creating executor via *New* to load batches with user defined type implementing *BatchUnit* interface
```go
// getBatchUnit returns CustomBatchUnit that implements BatchUnit interface
func getBatchUnit() batchproc.BatchUnit {
	return &CustomBatchUnit{}
}
```
- *New* also accepts an optional last parameter (batchSize) if user prefers to provide any preferred batch size

##### 2. **BatchUnit**
- It's an interface defining two methods Compute & GetResult
```go
type BatchUnit interface {
	Compute(int, int, interface{}) error
	GetResult() interface{}
}

```
- This provides a way for user to have customize batch steps as per use-case
```go
// CustomBatchUnit : define your own struct type
type CustomBatchUnit struct {
	result int
}

// Compute : define custom implementation for this method as per use case & store result in struct attribute
func (t *CustomBatchUnit) Compute(start, end int, collection interface{}) error {
	if data, ok := collection.([]int); ok {
		data = data[start:end]
		for _, value := range data {
			t.result += value
		}

		return nil
	}

	return fmt.Errorf("type assertion error")
}

// GetResult : return result attribute from defined struct type
func (t *CustomBatchUnit) GetResult() interface{} {
	return t.result
}
```

##### 3. **Run**
- Executes batch processing with loaded batch units
- Internally each batch will process under a separate rungroup (concurrent routine)
- Any single batch failure will interrrupt other batches & cancel overall batch processing
```go
// Computation stage
if err := executor.Run(); err != nil {
	log.Println(err)
	return
}
```

##### 4. **Aggregate**
- Aggregate performs final stage of the batch processing, where results from different batches are 
aggregated to form overall result.
- User should pass own aggregation function to process batch result as desired
```go
// Aggregation stage
result := executor.Aggregate(aggregation)
```
- **AggregationFunc** is a predefined type, to serve different case-to-case implementations `type AggregationFunc func([]interface{}) interface{}`
```go
// aggregation : treat results gathered from all batches as required
func aggregation(results []interface{}) interface{} {
	var res int

	for _, result := range results {
		if total, ok := result.(int); ok {
			res += total
		}
	}

	return res
}
```

### :rocket: Benchmark results

```
goos: linux
goarch: amd64
pkg: github.com/aryannr97/batchproc
cpu: Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
```

|                           Test case                               | b.N iterations | Avg processsing time |
| ----------------------------------------------------------------- | -------------- | -------------------- |
| Addition_of_first_10_million_integers_without_batch_processing-2  |      100       |    10665257 ns/op    |
| Addition_of_first_10_million_integers_with_batch_processing-2     |      251       |     4620470 ns/op    |


```
goos: linux
goarch: amd64
pkg: github.com/aryannr97/batchproc
cpu: Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz
```
|                           Test case                               | b.N iterations | Avg processsing time |
| ----------------------------------------------------------------- | -------------- | -------------------- |
| Addition_of_first_10_million_integers_without_batch_processing-12 |      184       |     6394545 ns/op    |
| Addition_of_first_10_million_integers_with_batch_processing-12    |      295       |     4074257 ns/op    |

As observed, batch processing can nearly run for **twice the number of iterations** and **reduce processing time by half**

### :open_file_folder: Example
To execute the demo program, run command `make demo`


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## v1.0.0
### Changed
 - Batch processor implementation
 - Updated badges
 - Updated README
 - Auto release actions

test_config.go
package batchproc

// testRunKey is user defined type for unit test key to be added in request context
type testRunKey string

// testRun defines constant key for unit test
const TestRun testRunKey = "testRun"


batchproc_test.go
package batchproc

import (
	"context"
	"reflect"
	"testing"
)

func TestNew(t *testing.T) {
	ctx := context.Background()
	newCtx := context.WithValue(ctx, TestRun, true)
	type args struct {
		ctx           context.Context
		id            string
		totalCount    int
		data          interface{}
		loadBatches   GetBatchUnitFunc
		batchSize     []int
		testDataCount int
	}
	tests := []struct {
		name string
		args args
		want int
	}{
		{
			"Get new instance of batch process executor",
			args{
				ctx,
				"unit-test",
				0,
				nil,
				getValidBatchUnit,
				[]int{},
				2000,
			},
			16,
		},
		{
			"Get new instance of batch process executor with fixed batchSize",
			args{
				ctx,
				"unit-test",
				0,
				nil,
				getValidBatchUnit,
				[]int{100},
				2000,
			},
			20,
		},
		{
			"Get new instance of batch process executor with min number of batches",
			args{
				ctx,
				"unit-test",
				0,
				nil,
				getValidBatchUnit,
				[]int{},
				100,
			},
			1,
		},
		{
			"Get new instance of batch process executor with min number of batches",
			args{
				ctx,
				"unit-test",
				0,
				nil,
				getValidBatchUnit,
				[]int{},
				100,
			},
			1,
		},
		{
			"Get new instance of batch process executor with 4 number of batches",
			args{
				ctx,
				"unit-test",
				0,
				nil,
				getValidBatchUnit,
				[]int{},
				500,
			},
			4,
		},
		{
			"Get new instance of batch process executor with 8 number of batches",
			args{
				ctx,
				"unit-test",
				0,
				nil,
				getValidBatchUnit,
				[]int{},
				1000,
			},
			8,
		},
		{
			"Get new instance of batch process executor with 16 number of batches",
			args{
				ctx,
				"unit-test",
				0,
				nil,
				getValidBatchUnit,
				[]int{},
				2000,
			},
			16,
		},
		{
			"Get new instance of batch process executor with max number of batches",
			args{
				ctx,
				"unit-test",
				0,
				nil,
				getValidBatchUnit,
				[]int{},
				2500,
			},
			20,
		},
		{
			"Get new instance of batch process executor with modified test context",
			args{
				newCtx,
				"unit-test",
				0,
				nil,
				getValidBatchUnit,
				[]int{},
				5,
			},
			4,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			testCollection := prepareTestCollection(tt.args.testDataCount)
			tt.args.totalCount = len(testCollection)
			tt.args.data = testCollection
			if got := New(tt.args.ctx, tt.args.id, tt.args.totalCount, tt.args.data, tt.args.loadBatches, tt.args.batchSize...); got.GetBatchCount() != tt.want {
				t.Errorf("batchCount = %v want = %v", got.GetBatchCount(), tt.want)
			}
		})
	}
}

func TestExecutor_Run(t *testing.T) {
	ctx := context.Background()
	testCollection := prepareTestCollection(2000)

	type args struct {
		ctx         context.Context
		id          string
		totalCount  int
		data        interface{}
		loadBatches GetBatchUnitFunc
		batchSize   []int
	}
	tests := []struct {
		name    string
		args    args
		wantErr bool
	}{
		{
			"Successful batch processing",
			args{
				ctx,
				"unit-test",
				len(testCollection),
				testCollection,
				getValidBatchUnit,
				[]int{},
			},
			false,
		},
		{
			"Failed batch processing",
			args{
				ctx,
				"unit-test",
				len(testCollection),
				testCollection,
				getErrorBatchUnit,
				[]int{},
			},
			true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			e := New(tt.args.ctx, tt.args.id, tt.args.totalCount, tt.args.data, tt.args.loadBatches, tt.args.batchSize...)
			if err := e.Run(); (err != nil) != tt.wantErr {
				t.Errorf("Executor.Run() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestExecutor_Aggregate(t *testing.T) {
	ctx := context.Background()
	testCollection := prepareTestCollection(2000)

	type fields struct {
		ctx         context.Context
		id          string
		totalCount  int
		data        interface{}
		loadBatches GetBatchUnitFunc
		batchSize   []int
	}
	type args struct {
		aggregation AggregationFunc
	}
	tests := []struct {
		name   string
		fields fields
		args   args
		want   interface{}
	}{
		{
			"Aggregate batch results to get final sum of first 2000 +ve integers",
			fields{
				ctx,
				"unit-test",
				len(testCollection),
				testCollection,
				getValidBatchUnit,
				[]int{},
			},
			args{
				aggregation,
			},
			2001000,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			e := New(tt.fields.ctx, tt.fields.id, tt.fields.totalCount, tt.fields.data, tt.fields.loadBatches, tt.fields.batchSize...)
			if err := e.Run(); err != nil {
				t.Errorf("Executor.Run() = %v", err)
			}
			if got := e.Aggregate(tt.args.aggregation); !reflect.DeepEqual(got, tt.want) {
				t.Errorf("Executor.Aggregate() = %v, want %v", got, tt.want)
			}
		})
	}
}

func BenchmarkBatchprocessor(b *testing.B) {
	ctx := context.Background()
	testCollection := prepareTestCollection(tenMillion)

	type args struct {
		ctx         context.Context
		id          string
		totalCount  int
		data        interface{}
		loadBatches GetBatchUnitFunc
		batchSize   []int
		batchFlag   bool
	}
	tests := []struct {
		name string
		args args
	}{
		{
			"Addition of first 10 million integers without batch processing",
			args{
				ctx,
				"unit-test",
				len(testCollection),
				testCollection,
				getValidBatchUnit,
				[]int{},
				false,
			},
		},
		{
			"Addition of first 10 million integers with batch processing",
			args{
				ctx,
				"unit-test",
				len(testCollection),
				testCollection,
				getValidBatchUnit,
				[]int{},
				true,
			},
		},
	}

	for _, tt := range tests {
		b.Run(tt.name, func(b *testing.B) {
			if tt.args.batchFlag {
				e := New(tt.args.ctx, tt.args.id, tt.args.totalCount, tt.args.data, tt.args.loadBatches, tt.args.batchSize...)
				for n := 0; n < b.N; n++ {
					_ = e.Run()
					e.Aggregate(aggregation)
				}
			} else {
				e := &TestBatchUnit{}
				for n := 0; n < b.N; n++ {
					_ = e.Compute(0, tt.args.totalCount, tt.args.data)
				}
			}
		})
	}
}


batchproc_mockdata_test.go
package batchproc

import (
	"fmt"
)

const (
	tenMillion = 10000000
)

var (
	testCollection = []int{}
)

// TestBatchUnit is custom type implementing BatchUnit interface
type TestBatchUnit struct {
	result int
}

// Compute performs addition portion of given slice of numbers
func (t *TestBatchUnit) Compute(start, end int, collection interface{}) error {
	if data, ok := collection.([]int); ok {
		data = data[start:end]
		count := 0
		for _, value := range data {
			count += value
		}
		t.result += count
	}

	return nil
}

// GetResult returns result stored during Compute operation
func (t *TestBatchUnit) GetResult() interface{} {
	return t.result
}

// // TestBatchUnit is custom type implementing BatchUnit interface
type TestNegativeBatchUnit struct{}

// Compute intentionally returns error for unit testing
func (t *TestNegativeBatchUnit) Compute(start, end int, data interface{}) error {
	return fmt.Errorf("test batch compute error")
}

// GetResult returns result stored during Compute operation
func (t *TestNegativeBatchUnit) GetResult() interface{} {
	return nil
}

// getValidBatchUnit returns instance of TestBatchUnit
func getValidBatchUnit() BatchUnit {
	return &TestBatchUnit{}
}

// getErrorBatchUnit returns instance of TestNegativeBatchUnit
func getErrorBatchUnit() BatchUnit {
	return &TestNegativeBatchUnit{}

}

// prepareTestCollection prepares slice of given N number of integers
func prepareTestCollection(N int) []int {
	testCollection = []int{}
	for i := 1; i <= N; i++ {
		testCollection = append(testCollection, i)
	}

	return testCollection
}

// aggregation is user defined custom AggregationFunc, which adds integer result of all batches
func aggregation(results []interface{}) interface{} {
	var res int

	for _, result := range results {
		if total, ok := result.(int); ok {
			res += total
		}
	}

	return res
}


cmd/demo.go
package main

import (
	"context"
	"fmt"
	"log"

	"github.com/aryannr97/batchproc"
)

func main() {
	data := getData()

	// Creation stage
	executor := batchproc.New(context.Background(), "main", len(data), data, getBatchUnit)
	fmt.Printf("Created batch executor loaded with number of batches: %v\n", executor.GetBatchCount())

	// Computation stage
	fmt.Println("Batch processor execution starting")
	if err := executor.Run(); err != nil {
		log.Println(err)
		return
	}
	// Aggregation stage
	fmt.Println("Batch processor result aggregation starting")
	result := executor.Aggregate(aggregation)

	fmt.Printf("Addition of first 2000 integers is: %v\n", result)
	fmt.Printf("Batch processor took %vns for execution\n", executor.ElapsedDuration.Nanoseconds())

}

func getData() []int {
	data := []int{}
	for i := 1; i <= 2000; i++ {
		data = append(data, i)
	}

	return data
}

// CustomBatchUnit : define your own struct type
type CustomBatchUnit struct {
	result int
}

// Compute : define custom implementation for this method as per use case & store result in struct attribute
func (t *CustomBatchUnit) Compute(start, end int, collection interface{}) error {
	if data, ok := collection.([]int); ok {
		data = data[start:end]
		count := 0
		for _, value := range data {
			count += value
		}

		t.result += count

		return nil
	}

	return fmt.Errorf("type assertion error")
}

// GetResult : return result attribute from defined struct type
func (t *CustomBatchUnit) GetResult() interface{} {
	return t.result
}

// getBatchUnit returns CustomBatchUnit that implements BatchUnit interface
func getBatchUnit() batchproc.BatchUnit {
	return &CustomBatchUnit{}
}

// aggregation : treat results gathered from all batches as required
func aggregation(results []interface{}) interface{} {
	var res int

	for _, result := range results {
		if total, ok := result.(int); ok {
			res += total
		}
	}

	return res
}


batchproc.go
package batchproc

import (
	"context"
	"fmt"
	"time"

	"github.com/bharat-rajani/rungroup"
)

// New creates, initializes & returns a new batch Executor,
//
// id is used to mark each batch with unique identification e.g {id}-batch-0,
//
// totalCount is the total size of the collection,
//
// data is actual collection need to be processed in batched fashion,
//
// getBatchUnit is used to fetch user defined type implementing BatchUnit interface and load batches,
//
// batchSize is optional & can be passed to override dynamic batch sizing.
func New(ctx context.Context, id string, totalCount int, data interface{}, getBatchUnit GetBatchUnitFunc, batchSize ...int) *Executor {
	e := &Executor{
		ctx:        ctx,
		id:         id,
		totalCount: totalCount,
		collection: data,
	}

	if len(batchSize) == 0 {
		e.prepareDynamicLoading()
	} else {
		e.prepareFixedLoading(batchSize[0])
	}

	// load batch units with user defined func
	for i := 0; i < e.batchCount; i++ {
		e.batches = append(e.batches, getBatchUnit())
	}

	return e
}

// Executor represents concurrent batch processing entity
type Executor struct {
	ctx             context.Context
	id              string
	batchSize       int
	totalCount      int
	batchCount      int
	batches         []BatchUnit
	collection      interface{}
	startTime       time.Time
	endTime         time.Time
	ElapsedDuration time.Duration
}

// prepareDynamicLoading loads number of batches, batchSize depending on totalCount,
// number of batches formed will be min:1 max:20.
func (e *Executor) prepareDynamicLoading() {
	/**
	Scale down the sample space by a factor of 100 if function is triggered through unit tests.
	Testing different number of batches can be accomplished even with smaller size of mockData.
	*/
	scaleDownFactor := 1
	if test, _ := e.ctx.Value(TestRun).(bool); test {
		scaleDownFactor = 100
	}

	if e.totalCount <= (100 / scaleDownFactor) {
		e.batchCount = 1
	} else if e.totalCount <= (500 / scaleDownFactor) {
		e.batchCount = 4
	} else if e.totalCount <= (1000 / scaleDownFactor) {
		e.batchCount = 8
	} else if e.totalCount <= (2000 / scaleDownFactor) {
		e.batchCount = 16
	} else {
		e.batchCount = 20
	}

	e.batchSize = e.totalCount / e.batchCount
}

// prepareFixedLoading loads number of batches based on fixed batchSize
func (e *Executor) prepareFixedLoading(batchSize int) {
	e.batchSize = batchSize
	e.batchCount = e.totalCount / e.batchSize
}

// Run triggers concurrent routines for computation through batch units
func (e *Executor) Run() error {
	e.startTime = time.Now()
	countRg, ctxRg := rungroup.WithContext(e.ctx)

	for i, item := range e.batches {
		index := i
		batch := item
		id := fmt.Sprintf("%v-batch-%v", e.id, index)

		countRg.GoWithFunc(func(context.Context) error {
			start := index * e.batchSize
			end := (index * e.batchSize) + e.batchSize
			// Consider all the remaining items for last batch
			if index == e.batchCount-1 {
				end = e.totalCount
			}
			// Perform computation per batch
			err := batch.Compute(start, end, e.collection)
			return err
		}, ctxRg, true, id)
	}

	// Wait (blocking) till all the batches are done processing
	if countRgErr := countRg.Wait(); countRgErr != nil {
		return countRgErr
	}

	return nil
}

// Aggregate is used for result aggregation and mark closure of batch processing,
// user should define own aggregation func to process results from different batches.
func (e *Executor) Aggregate(aggregation AggregationFunc) interface{} {
	results := []interface{}{}
	for _, batch := range e.batches {
		results = append(results, batch.GetResult())
	}

	response := aggregation(results)

	// Mark process completion post result aggregation
	e.onComplete()

	return response
}

// onComplete performs logging,reporting activities on batch processing completion
func (e *Executor) onComplete() {
	e.endTime = time.Now()
	e.ElapsedDuration = time.Duration(e.endTime.Sub(e.startTime))
}

// GetBatchCount returns number of batches for given executor
func (e *Executor) GetBatchCount() int {
	return e.batchCount
}
